File: GD32驱动器/GD32FirmwareUpdater/src/GD32FirmwareUpdater/Protocol/CrcCalculator.cs
// GD32低压驱动器Boot升级协议 - CRC16校验算法
// 采用CRC16/Modbus算法，多项式0xA001，初始值0xFFFF

namespace GD32FirmwareUpdater.Protocol
{
    /// <summary>
    /// CRC16校验计算器（Modbus标准）
    /// 算法说明：
    /// 1. 初始CRC值为0xFFFF
    /// 2. 逐字节与CRC异或
    /// 3. 对每一位进行移位和多项式判断
    /// 4. 输出2字节校验值（小端模式存储）
    /// </summary>
    public static class CrcCalculator
    {
        // CRC16/Modbus 多项式（反转形式）
        private const ushort POLYNOMIAL = 0xA001;

        // CRC初始值
        private const ushort INITIAL_VALUE = 0xFFFF;

        /// <summary>
        /// 计算给定数据的CRC16校验值
        /// </summary>
        /// <param name="data">待校验的数据字节数组</param>
        /// <returns>CRC16校验值</returns>
        public static ushort Calculate(byte[] data)
        {
            return Calculate(data, 0, data.Length);
        }

        /// <summary>
        /// 计算指定范围数据的CRC16校验值
        /// </summary>
        /// <param name="data">数据字节数组</param>
        /// <param name="offset">起始偏移</param>
        /// <param name="length">计算长度</param>
        /// <returns>CRC16校验值</returns>
        public static ushort Calculate(byte[] data, int offset, int length)
        {
            ushort crc = INITIAL_VALUE;

            for (int i = offset; i < offset + length; i++)
            {
                // 当前字节与CRC低字节异或
                crc ^= data[i];

                // 对8个bit逐位处理
                for (int j = 0; j < 8; j++)
                {
                    if ((crc & 0x0001) != 0)
                    {
                        // 最低位为1：右移一位后与多项式异或
                        crc = (ushort)((crc >> 1) ^ POLYNOMIAL);
                    }
                    else
                    {
                        // 最低位为0：仅右移一位
                        crc = (ushort)(crc >> 1);
                    }
                }
            }

            return crc;
        }

        /// <summary>
        /// 将CRC值转为小端字节数组（低字节在前）
        /// </summary>
        public static byte[] ToLittleEndianBytes(ushort crc)
        {
            return new byte[]
            {
                (byte)(crc & 0xFF),        // 低字节
                (byte)((crc >> 8) & 0xFF)  // 高字节
            };
        }
    }
}

File: GD32驱动器/GD32FirmwareUpdaterQt/src/protocol/CrcCalculator.cpp
// ============================================================
//  CRC16 校验计算器实现 — CRC16/Modbus
//
//  【Qt知识点】QByteArray 是字节容器（类似 C# 的 byte[]），
//  at(i) 返回 char（可能有符号），因此用 static_cast<quint8>
//  保证按无符号字节参与位运算 —— 这是 C/C++ 位运算的常见坑，
//  C# 的 byte 天然无符号，C++ 的 char 则可能为有符号。
// ============================================================
#include "CrcCalculator.h"

quint16 CrcCalculator::calculate(const QByteArray &data)
{
    return calculate(data, 0, data.size());
}

quint16 CrcCalculator::calculate(const QByteArray &data, int offset, int length)
{
    // 边界保护：越界时自动裁剪到合法范围
    if (offset < 0 || offset >= data.size())
        return INITIAL_VALUE;
    if (length < 0 || offset + length > data.size())
        length = data.size() - offset;

    quint16 crc = INITIAL_VALUE;

    for (int i = offset; i < offset + length; ++i)
    {
        // 当前字节与 CRC 低字节异或（关键：无符号转换）
        crc ^= static_cast<quint8>(data.at(i));

        // 对 8 个 bit 逐位处理
        for (int j = 0; j < 8; ++j)
        {
            if (crc & 0x0001)
                crc = static_cast<quint16>((crc >> 1) ^ POLYNOMIAL);  // 最低位为1：右移异或多项式
            else
                crc = static_cast<quint16>(crc >> 1);                 // 最低位为0：仅右移
        }
    }

    return crc;
}

QByteArray CrcCalculator::toLittleEndianBytes(quint16 crc)
{
    QByteArray bytes;
    bytes.append(static_cast<char>(crc & 0xFF));         // 低字节在前
    bytes.append(static_cast<char>((crc >> 8) & 0xFF));  // 高字节在后
    return bytes;
}
